Cli patch

bryantChhun opened the pull request:
Sync with master and..

Tested both `extract_patches` and `build_trajectories` using a subset of data from microglia experiments.  They both receive and generate files well.

There are very few major changes in this except to adjust function input/outputs.

smguo reviewed (changes requested):
Same as #27. Can you merge the master to this branch? I'll take another look after that. Thanks!

bryantChhun commented:
Ok conflicts should be resolved.

Additionally, i added a config file that references data on ESS.  You can use this config file to verify the patch pipeline yourself:

pycharm config
```
# run_patch, extract_patches
python run_patch.py -m "extract_patches" -c ./configs/bryant_rubella_experiments.yml
```
The above should pull `.npy`, and `_NNProbabilities.npy` from:
`/gpfs/CompMicro/projects/dynamorph/microglia/rubella_experiments/NOVEMBER/new-3-17-2021/seg_test/raw`
and `.cell_pixel_assignments.pkl`, and `cell_positions.pkl` from: 
`/gpfs/CompMicro/projects/dynamorph/microglia/rubella_experiments/NOVEMBER/new-3-17-2021/seg_test/supp`

Then write `patch_t-<timepointid>-<cellID>.jpg`, `stacks_<timepoint>.pkl`, and updated `cell_positions.pkl` to
`/gpfs/CompMicro/projects/dynamorph/microglia/rubella_experiments/NOVEMBER/new-3-17-2021/seg_test/supp`

smguo commented on run_patch.py:
```diff
@@ -67,7 +70,7 @@ def main(arguments_):
         _sites = segment_sites[sep[i]:sep[i + 1]]
-        args = (raw, supp, channels, None, _sites)
+        args = (raw, supp, _sites, config_)
         p = Worker(args, gpuid=i, method=method)
```
Both  'extract_patches' and 'build_trajectories' run on CPU and can really speed up using more than 4 workers. Maybe rename the parameter 'gpuid' to 'worker_id'

bryantChhun replied:
yes good, point, i can also update the config file to reflect this.

smguo commented on run_VAE.py:
```diff
@@ -1,8 +1,11 @@
-import os
+import os, sys
 import argparse
+from multiprocessing import Process
```
Curious about the reason to revert back to python `multiprocessing` from `torch.multiprocessing`. `torch.multiprocessing` is recommended in [pytorch doc](https://pytorch.org/docs/master/notes/multiprocessing.html#cuda-in-multiprocessing)

bryantChhun replied:
hm, i'm not sure i intended to change this in this PR.  This might reflect and older version of the script that hadn't been updated.  Will change it back.

smguo commented on pipeline/patch_VAE.py:
```diff
@@ -471,7 +499,9 @@ def process_VAE(summary_folder: str,
-            output = model(sample)[0]
+            # output = model(sample)[0]
+            # model still uses channels last?
+            output = model(torch.transpose(sample[0], 0, 1))[0]
```
the issue here is the model is 4D but the patches are 5D (with additional z dimension). You can reshape the tensor with `sample.reshape([-1, n_channels, x_size, y_size])`

bryantChhun replied:
This portion of `patch_VAE.py` is not used in this PR, so this block of code is possibly outdated.  I can match it with master, or we can implement the reshape as you suggest, then fix it in the next PR that focuses on `run_VAE.py`.  let me know what you think

smguo replied:
Sounds good. We can fix it in the next PR

smguo commented on pipeline/patch_VAE.py:
```diff
@@ -54,10 +51,18 @@ def extract_patches(summary_folder: str,
+    skip_boundary = config.patch.skip_boundary
 
     for site in sites:
-        site_path = os.path.join(summary_folder + '/' + site + '.npy')
```
docstrings need to be updated

smguo commented on pipeline/patch_VAE.py:
```diff
@@ -162,24 +171,24 @@ def assemble_VAE(summary_folder: str,
     dataset, fs = prepare_dataset_v2(dat_fs, cs=channels)
     assert fs == sorted(fs)
     
-    print(f"\tsaving {os.path.join(summary_folder, '%s_file_paths.pkl' % well)}")
```
docstrings need to be updated

smguo commented on pipeline/patch_VAE.py:
```diff
@@ -349,14 +359,10 @@ def import_object(module_name, obj_name, obj_type='class'):
         raise
 
-def process_VAE(summary_folder: str,
+def process_VAE(raw_folder: str,
```
docstrings need to be updated

smguo reviewed (approved):
@bryantChhun Looks good! I added some comments. Please take a look and address them as you see fit.
